src/pip/_internal/download.py

```diff
@@ -141,38 +147,141 @@ def user_agent():
+        if cred is not None:
+            logger.debug("Found credentials in keyring")
+            return cred.username, cred.password
+
+
 class MultiDomainBasicAuth(AuthBase):
 
-    def __init__(self, prompting=True):
-        # type: (bool) -> None
+    def __init__(self, prompting=True, index_urls=None):
+        # type: (bool, Optional[Values]) -> None
         self.prompting = prompting
+        self.index_urls = index_urls
         self.passwords = {}  # type: Dict[str, AuthInfo]
-
-    def __call__(self, req):
-        parsed = urllib_parse.urlparse(req.url)
-
-        # Split the credentials from the netloc.
-        netloc, url_user_password = split_auth_from_netloc(parsed.netloc)
-
-        # Set the url of the request to the url without any credentials
-        req.url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
+        # When the user is prompted to enter credentials and keyring is
+        # available, we will offer to save them. If the user accepts,
+        # this value is set to the credentials they entered. After the
+        # request authenticates, the caller should call
+        # ``save_credentials`` to save these.
+        self._credentials_to_save = None   # type: Tuple[str, str, str]
+
+    def _get_index_url(self, url):
+        """Returns the original index URL matching the requested URL.
+
+        Cached or dynamically generated credentials may work against
+        the original index URL rather than just the netloc.
+
+        The provided url should have had its username and password
+        removed already. If the original index url had credentials then
+        they will be included in the return value.
+
+        Returns None if no matching index was found, or if --no-index
+        was specified by the user.
+        """
+        if not url or not self.index_urls:
+            return None
+
+        for u in self.index_urls:
+            prefix = remove_auth_from_url(u).rstrip("/") + "/"
+            if url.startswith(prefix):
+                return u
+
+    def _get_url_and_credentials(self, original_url,
+                                 allow_netrc=True, allow_keyring=True):
+        """Returns the credentials to use for the provided URL.
+
+        If allowed, netrc and keyring may be used to obtain the
+        correct credentials.
+
+        Returns (url_without_credentials, username, password). Note
+        that even if the original URL contains credentials, the result
+        of this function may include different values.
+        """
+
+        # Split the credentials and netloc from the url.
+        url, netloc, url_user_password = split_auth_netloc_from_url(
+            original_url)
 
         # Use any stored credentials that we have for this netloc
         username, password = self.passwords.get(netloc, (None, None))
 
+        # Early exit if we have cached credentials for this netloc
+        if username is not None and password is not None:
```

Comment:
Here you are requiring both of them to be non-`None` in order to use them, but when the cache is populated below, only one is required to be truthy. It seems like the condition for deciding when to store and when to use the credentials should maybe be the same. It would be worth getting clarity on this since there are similar checks sprinkled throughout this method.

Reply:
Agreed. I think the best option is to require non-`None` throughout, since an empty password is technically valid (and the helper functions we already have will return `user, None` for `user@netloc` but `user, ""` for `user:@netloc`)